### PR TITLE
prime the serializer in a helper thread to cache reflection calls mak…

### PR DESCRIPTION
**Pull request by thomaszurkan-optimizely**
…ing subsequent calls to create and serialize EventBatch must faster.

This reduces activate calls on Android from 100 ms to 15 ms.

### CODE REVIEW

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java [hunk 1]
```diff
@@ -61,8 +63,32 @@ public EventBuilder(EventBatch.ClientEngine clientEngine, String clientVersion)
...
+                primeSerializer();
+            }
+        };
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        executor.submit(r);
+
     }
 
+    private void primeSerializer() {
```
Comment (mikeng13) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: Can you add a comment explaining why this is necessary?

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java [hunk 2]
```diff
@@ -61,8 +63,32 @@ public EventBuilder(EventBatch.ClientEngine clientEngine, String clientVersion)
...
+        Decision decision = new Decision("", "",
```
Comment (mikeng13) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: I'd like to be on the safer side and put this inside a `try/catch`. Worst that can happen then is that the priming fails and the very first API call that sends events will take longer

Comment (thomaszurkan-optimizely) [reply core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: Because it is in its own Runnable thread, worst case is thread dies.

**Comment by jophde** [comment]
Do we have any benchmarks around how expensive of call this is caching?

**Comment by thomaszurkan-optimizely** [comment]
@jophde I wrapped some times around things and discovered that the creation of the EventBatch and its associated objects made up about 60% of the time while the serializer took up another 40% of a total somewhere over 100 ms.  It now sub 100 ms.

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java [hunk 3]
```diff
@@ -61,21 +67,77 @@ public EventBuilder(EventBatch.ClientEngine clientEngine, String clientVersion)
...
     }
 
+    private void waitForPrimedEvent() {
+        if (primedEvent != null) {
+            logger.debug("had to wait for prime");
+            try {
+                primedEvent.get();
+            }
+            catch (Exception e) {
+                logger.error("Problem getting lock ", e);
```
Comment (mikeng13) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: Can you give a bit context in this message so the user will have an easier time figuring what's wrong?

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java [hunk 4]
```diff
@@ -61,21 +67,77 @@ public EventBuilder(EventBatch.ClientEngine clientEngine, String clientVersion)
         this.clientEngine = clientEngine;
         this.clientVersion = clientVersion;
         this.serializer = DefaultJsonSerializer.getInstance();
+
+        Callable<EventBatch> callable = new Callable<EventBatch>() {
+            @Override
+            public EventBatch call() throws Exception {
+                EventBatch eventBatch = null;
+                try {
+                    eventBatch = primeSerializer();
```
Comment (mikeng13) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: Should this try to return?

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java [hunk 5]
```diff
@@ -61,21 +67,77 @@ public EventBuilder(EventBatch.ClientEngine clientEngine, String clientVersion)
...
+                "", false);
+        Event impressionEvent = new Event(System.currentTimeMillis(),UUID.randomUUID().toString(), "",
+                ACTIVATE_EVENT_KEY, null, null, null, ACTIVATE_EVENT_KEY, null);
+
+        Snapshot snapshot = new Snapshot(Arrays.asList(decision), Arrays.asList(impressionEvent));
+        Visitor visitor = new Visitor("", null, new ArrayList<Attribute>(), Arrays.asList(snapshot));
+        List<Visitor> visitors = Arrays.asList(visitor);
+        EventBatch eventBatch = new EventBatch(clientEngine.getClientEngineValue(), clientVersion, "", visitors, true, "", "");
+        String payload = this.serializer.serialize(eventBatch);
+        logger.debug("JSON Serializer primed for payload {}", payload);
```
Comment (mikeng13) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: Nit: ... for event payload

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java [hunk 6]
```diff
@@ -61,21 +67,77 @@ public EventBuilder(EventBatch.ClientEngine clientEngine, String clientVersion)
...
+
+        return  eventBatch;
     }
 
+    private void waitForPrimedEvent() {
+        if (primedEvent != null) {
+            logger.debug("had to wait for prime");
```
Comment (mikeng13) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: This message could be a bit more descriptive

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java [hunk 7]
```diff
@@ -61,21 +67,77 @@ public EventBuilder(EventBatch.ClientEngine clientEngine, String clientVersion)
...
         this.serializer = DefaultJsonSerializer.getInstance();
+
+        Callable<EventBatch> callable = new Callable<EventBatch>() {
+            @Override
+            public EventBatch call() throws Exception {
+                EventBatch eventBatch = null;
+                try {
+                    eventBatch = primeSerializer();
+                } catch (Exception e) {
+                    logger.error("Problem priming the JsonSerializer ", e);
```
Comment (mikeng13) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventBuilder.java]: This error message is too generic. Let's give the user something more specific

**Comment by thomaszurkan-optimizely** [comment]
I want to show one other option in a different pr

**Pull request closed without merging**